res/ari/ari_websocket_requests.c

```diff
@@ -0,0 +1,312 @@
+/*
+ * Asterisk -- An open source telephony toolkit.
+ *
+ * Copyright (C) 2025, Sangoma Technologies Corporation
+ *
+ * George Joseph <gjoseph@sangoma.com>
+ *
+ * See http://www.asterisk.org for more information about
+ * the Asterisk project. Please do not directly contact
+ * any of the maintainers of this project for assistance;
+ * the project provides a web site, mailing lists and IRC
+ * channels for your use.
+ *
+ * This program is free software, distributed under the terms of
+ * the GNU General Public License Version 2. See the LICENSE file
+ * at the top of the source tree.
+ */
+
+#include "asterisk.h"
+
+#include "ari_websockets.h"
+#include "asterisk/ari.h"
+#include "asterisk/json.h"
+#include "asterisk/stasis_app.h"
+
+struct rest_request_msg {
+	char *request_type;
+	char *transaction_id;
+	char *request_id;
+	enum ast_http_method method;
+	char *uri;
+	char *content_type;
+	struct ast_variable *query_strings;
+	struct ast_json *body;
+};
+
+static void request_destroy(struct rest_request_msg *request)
+{
+	if (!request) {
+		return;
+	}
+
+	ast_free(request->request_type);
+	ast_free(request->transaction_id);
+	ast_free(request->request_id);
+	ast_free(request->uri);
+	ast_free(request->content_type);
+	ast_variables_destroy(request->query_strings);
+	ast_json_unref(request->body);
+
+	ast_free(request);
+}
+
+#define SET_RESPONSE_AND_EXIT(_reponse_code, _reponse_text, \
+	_reponse_msg, _remote_addr, _request, _request_msg) \
+({ \
+	RAII_VAR(char *, _msg_str, NULL, ast_json_free); \
+	if (_request_msg) { \
+		_msg_str = ast_json_dump_string_format(_request_msg, AST_JSON_COMPACT); \
+		if (!_msg_str) { \
+			response->response_code = 500; \
+			response->response_text = "Server error.  Out of memory"; \
+		} \
+	} \
+	response->message = ast_json_pack("{ s:s }", \
+		"message", _reponse_msg); \
+	response->response_code = _reponse_code; \
+	response->response_text = _reponse_text; \
+	SCOPE_EXIT_LOG_RTN_VALUE(_request, LOG_WARNING, \
+		"%s: %s Request: %s\n", _remote_addr, _reponse_text, S_OR(_msg_str, "<none>")); \
+})
+
+static struct rest_request_msg *parse_rest_request_msg(
+	const char *remote_addr, struct ast_json *request_msg,
+	struct ast_ari_response *response)
+{
+	struct rest_request_msg *request = NULL;
+	RAII_VAR(char *, body, NULL, ast_free);
+	enum ast_json_nvp_ast_vars_code nvp_code;
+	char *query_string_start = NULL;
+	SCOPE_ENTER(4, "%s: Parsing RESTRequest message\n", remote_addr);
+
+	response->response_code = 200;
+	response->response_text = "OK";
+
+	if (!request_msg) {
+		SET_RESPONSE_AND_EXIT(500,
+			"Server error","No message to parse.",
+			remote_addr, request, NULL);
+	}
+
+	request = ast_calloc(1, sizeof(*request));
+	if (!request) {
+		SET_RESPONSE_AND_EXIT(500,
+			"Server error","Out of memory",
+			remote_addr, request, NULL);
+	}
+
+	/* transaction_id is optional */
```

Comment:
If transaction_id exists in request_msg, but the ast_strdup fails, does this break an understanding/contract between the server and the client as to the existence/usage of it? (Same applies to others)